### ISSUE TITLE
[atari2600][question/feature] Any API to draw on the screen?

### ISSUE THREAD

**Issue by Fabrizio-Caruso** [cc65/cc65#815]
Coding for the Atari 2600 will always remain a challenge to the coder because of its inherent hardware limits.

Does CC65 provide any APIs to draw on the screen just a point  on the screen?

Coding in C could be made easier make it easier (Batari BASIC seems to achieve this by abstracting most of the low level coding). I realize that this would be a huge amount of work and I am not asking for this.
Maybe something could be done, to ease coding the Atari VCS in C for non-VCS geeks.

Is there for instance a simple way to just display "Hello World" in C, at a given position on the screen?
The "Hello World" example provided in CC65 does something else.

**Comment by oliverschmidt** [comment]
> Does CC65 provide any APIs to draw on the screen just a point on the screen?

No.

> Is there for instance a simple way to just display "Hello World" in C, at a given position on the screen

No.

> I realize that this would be a huge amount of work and I am not asking for this.

Don't get me wrong but given this explicit statement this PR isn't an enhancement request, so the only thing left is a question. And as I answered the actual questions I'm closing this PR. You're of course free to change your mind and ask for an enhancement.

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt I just added the "enhancement" tag to make it a polite feature request.

**Comment by oliverschmidt** [comment]
I see.

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt I am not requesting anything of the same size of Batari BASIC (which provides simple high level APIs to code the Atari 2600). 
I have mentioned BATARI Basic because it proves that new nice games can be coded with high level APIs despite the common belief that the only way to code the Atari 2600 is by coding the "beam".
It would be nice to have some abstraction of the "beam" in CC65 like drawing "fat pixels" on the playfiled.

**Comment by oliverschmidt** [comment]
The fact that you direct your recent comment to me seems to imply a misunderstanding: I'm just the maintainer of cc65. If someone contributes improved Atari 2600 support (and doesn't go against our coding/style rules) I'm happy to merge it. So your issue isn't me not wanting it but nobody contributing it.

**Comment by Fabrizio-Caruso** [comment]
Sure. I don't expect anything to be done on my request in the next weeks or ever. Everything here is done on a voluntary basis. I am willing to help (and I have with bug-fixes and tests) but this is beyond my possibilities. 
On the other hand I may come up with some little contribution (maybe for the Supervision).

I am not sure if feature requests are welcome. I am not even sure CC65 maintainers would be happy to accept entire target-specific libs. I have the impression that the maintainers want to keep CC65 a lean dev-kit (compared to Z88DK) where only some small libs are integrated.

**Comment by oliverschmidt** [comment]
> I am not even sure CC65 maintainers would be happy to accept entire target-specific libs. I have the impression that the maintainers want to keep CC65 a lean dev-kit (compared to Z88DK) where only some small libs are integrated. I may be wrong.

At least regarding my POV your impression is spot-on. cc65 isn't a container for all types of libraries for all types of purposes for a single target. However, if someone would e.g. come up with a subset of the TGI API like

- tgi_clear
- tgi_gotoxy
- tgi_setcolor
- tgi_setpixel
- tgi_line
- tgi_lineto
- tgi_getmaxx
- tgi_getmaxy
- tgi_getaspectratio

implemented by code that's just part of the C library without any TGI driver and without going through the TGI kernel I could imagine well to merge it. One could imagine to add dummy

- tgi_install
- tgi_init
- tgi_done
- tgi_uninstall

which do nothing at all. Then one could at least theoretically write TGI programs working on other "traditional" TGI supporting targets too.

Instead of emulating (a subset of) CONIO one could add

- tgi_outtext
- tgi_outtextxy

like the other functions not using the TGI infrastructure but implemented straight with target-specific code. One could think to accept some semantic differences to "full" TGI support, e.g. saying that clipping isn't implemented (meaning the user just needs to make sure that no operation hits the screen border).

And as addition (not as a replacement !) of the above one could add some target specific code like sprites or alike using

- tgi_ioctl

Additionally it might make sense to have

- waitvsync

implemented. And (again if desired without actual driver) one could implement 

- joy_count
- joy_read

and maybe complement it with dummies for

- joy_install
- joy_uninstall

Maybe I missing the point but to me it seems that a pretty lightweight approach to supporting a console-type target is possible while holding on to the idea of common cc65 cross target APIs.

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt, one important API or functionality that is missing in TGI for any target (except Atari Lynx where some sort of sprites are available) is something I call "software sprite" but I mean just a copy of a bit pattern with some AND/XOR/OR operation.

**Comment by Fabrizio-Caruso** [comment]
@polluks yes, but it provides high levels methods to draw on 2600's playfield and methods to control sprites (even multi-plexing is abstracted: you get 6 logical sprites instead of 2).

**Comment by polluks** [comment]
Indeed, kernel support was already on my todo list :-)

**Comment by Fabrizio-Caruso** [comment]
@polluks it seems that this year I will send a letter to you instead of Santa Claus!
Most of my hopes for enhancement are in your TODO list. :-)

**Comment by polluks** [comment]
@Fabrizio-Caruso Even if we have "batari C", we will never have conio, and therefore no portability.

**Comment by polluks** [comment]
TGI? I'm really sorry but 2600 is impossible, 7800 is possible.

**Comment by Fabrizio-Caruso** [comment]
A partial TGI implementation would still be possible, e.g., tgi_setpixel (to set a pixel on the playfield).
What I and probably @oliverschmidt mean by multi-target support is that the same APIs should be used whenever reasonably possible. A limited number of lean target-specific APIs would be tolerated (as this is already the case for some targets such sound functions for the atmos target).

Since for the Atari 2600 we have nothing, I would be happy to have just something.

**Comment by oliverschmidt** [comment]
> What I and probably @oliverschmidt mean by multi-target support is that the same APIs should be used whenever reasonably possible.

Your assumption is spot-on :-)

> A limited number of lean target-specific APIs would be tolerated (as this is already the case for some targets such sound functions for the atmos target).

I again fully agree. In general I believe that there's just no viable option for a common sound API. The hardware is too different, the performance is too critical and the resources are too limited to allow for a useful common abstraction.

**Comment by oliverschmidt** [comment]
> TGI? I'm really sorry but 2600 is impossible

Did you read what I wrote in https://github.com/cc65/cc65/issues/815#issuecomment-447566847? Did you notice that I were __not__ talking about a TGI driver?

**Comment by groessler** [comment]
> A partial TGI implementation would still be possible, e.g., tgi_setpixel (to set a pixel on the playfield).

I'm not sure. Remember, the 2600 only has 128 bytes of RAM AFAIK. Where should the pixel data be stored? IDK how the BASIC works. But for cc65 you probably need to create some generic kernel for display. Has the BASIC an equivalent of tgi_setpixel()? IMO the 2600 target is primarily targeted to ASM programs.

**Comment by Fabrizio-Caruso** [comment]
@groessler you are right about set_pixel!
Nevertheless Batari BASIC proves that nice games can be coded in a high level language.
Maybe TGI is indeed impossible. 
Playfield graphics could be handled by some pre-defined layouts in ROM memory.

**Comment by Fabrizio-Caruso** [comment]
@groessler Batari BASIC does provide very low resolution bitmap mode (32x11):
https://alienbill.com/2600/basic/downloads/batari_users_guide.htm

It provides high level apis for multiplexed multicolor sprites, high resolution images, scores, etc.

**Comment by groessler** [comment]
@Fabrizio-Caruso, I know your game and I can imagine that you could make use of such a 32x11 bitmap mode.

But consider it from the perspective of a general toolchain. Who else would use that mode? 1, 10, 100, 1000 persons? Is it worth to invest the time and effort to implement something like this? I don't see many (any?) requests for something like this. So, I'm afraid, if you want to have this you probably need to implement it yourself. I don't see anyone stepping up, shouting "here", and doing the work.

The 2600 is rather difficult to program. One has to "follow the beam" on the screen and update things at the correct time. I'm happy that we have general assembler support for this system at least, but there are still things missing. Like the cycle-counting of DASM for example.

**Comment by polluks** [comment]
@Fabrizio-Caruso Indeed, we could have a little bit more support...

**Comment by Fabrizio-Caruso** [comment]
@groessler Forget about my game, which can exist without an Atari 2600 version. I have already beaten the record. :-P
We know for sure that the Atari 2600 can be coded without following the beam by looking at the many games produced in Batari BASIC.
I am not saying someone should code a C equivalent of BATARI Basic. That would be a LOT of work.

I am just saying that a little more C support would make this target more usable by non-Assembly and non-VCS ninjas.

**Comment by groessler** [comment]
As I said, patches are welcome...
I personally am not interested in working on the 2600 at this time.

**Comment by GorillaSapiens** [comment]
> We know for sure that the Atari 2600 can be coded without following the beam by looking at the many games produced in Batari BASIC.
> I am not saying someone should code a C equivalent of BATARI Basic. That would be a LOT of work.

i would not say that "coded without following the beam" is true.  BATARI Basic provides a number of kernels.  all of the kernels work by "following the beam".  the programmer gets control only during the brief period of time between the bottom of the screen and the vertical sync signal that restarts a frame.  the programmer controls the image by changing parameters in the kernel.  the kernel is tightly tied to the timing of the beam.  programmers write kernels in assembly, and keep track of how many cycles it takes for each opcode.  it is not a trivial task.  it is not a bitmapped display.  to satisfy the ask here, you really would wind up writing the C equivalent of BATARI Basic.

given the unique challenges of this platform, i don't think you're ever going to see "an API to just draw a point on the screen".

**Comment by polluks** [comment]
Indeed, cc65 should support some kind of kernel...
